src/test/java/com/blog/stream/pagination/PageSpliteratorTest.java

```diff
@@ -1,31 +1,284 @@
+
+        PageSpliterator<PagedResult<TestObject>, TestObject> spliterator = PageSpliterator.create(1000, 10, testObjectPagedRepository::fetchPage, PagedResult::getItems);
+        Set<Thread> threads = Sets.newHashSet();
+
+        ForkJoinPool pool = new ForkJoinPool(100);
+
+        ConcurrentArrayList<TestObject> allResultsc = new ConcurrentArrayList<>();
+
+        Stream<TestObject> peekProcessor = spliterator.stream().parallel()
+                .peek(user -> threads.add(Thread.currentThread()));
+
+        pool.submit(() -> peekProcessor.forEach(allResultsc::add));
+        pool.shutdown();
+        Assert.assertTrue(pool.awaitTermination(10, TimeUnit.SECONDS));
+
+        System.out.println("Concurrency: " + threads.size());
+        System.out.println("avail processors: " + Runtime.getRuntime().availableProcessors());
+
+        List<TestObject> allResults = new ArrayList<>();
+
+        allResultsc.iterator().forEachRemaining(allResults::add);
+        assertAllIdsMatch(allResults, testObjectPagedRepository.source);
+        Assert.assertTrue(threads.size() > 1);
+
+    }
 
     @Test
-    public void trySplit_ReturnsChildWithCurrentPageAndMovesOnToNextPage() throws Exception {
-        PageSpliterator<String> spliterator = new PageSpliterator<>(3, 100, 10, null);
+    public void getPagedStreamParallelShouldStreamThroughAllPagesUpToKnownThreads() {
+
+        PageSpliterator<PagedResult<TestObject>, TestObject> spliterator = PageSpliterator.create(1000, 100, testObjectPagedRepository::fetchPage, PagedResult::getItems);
+        Set<Thread> threads = Sets.newHashSet();
+
+        List<TestObject> allResults = spliterator.stream()
+                .parallel()
+                .peek(user -> threads.add(Thread.currentThread()))
+       .collect(Collectors.toList());
+
+        System.out.println("Concurrency: " + threads.size());
+        System.out.println("avail processors: " + Runtime.getRuntime().availableProcessors());
+
+        assertAllIdsMatch(source, allResults);
+
+        Assert.assertTrue(threads.size() > 1);
+
+    }
+
+    @Test
+    public void getPagedStreamParallelWithPrefetchShouldStreamThroughAllPagesUpToKnownThreads() {
+
+        doTest("count > page size, parallel", 100, 1000, true);
+        doTest("count = page size, parallel",100, 100, true);
+        doTest("count on less than page size, parallel",100, 99, true);
+        doTest("count < page size, parallel",100, 10, true);
+        doTest("count = 0, parallel",100, 0, true);
+
+        doTest("count > page size, sequential", 100, 1000, false);
+        doTest("count = page size, sequential",100, 100,false);
+        doTest("count on less than page size, sequential",100, 99,false);
+        doTest("count < page size, sequential",100, 10, false);
+        doTest("count = 0, sequential",100, 0, false);
+    }
+
+    private void doTest(String title, int pageSize, int totalResults, boolean isParallel) {
```

Comment:
I generally avoid using DRY for testing, aiming for less clicking around. Makes diagnosing problems easier and faster.

Ive been using DAMP as a standard for the tests in this repo

Reply:
The intent here was essentially a test template, but crappy name